libraries/AC_SafeRTL/AC_SafeRTL.cpp

```diff
@@ -0,0 +1,573 @@
+*    triggered, a more thorough cleanup occurs, and then the resulting path can
+*    is fed into WP_NAV to fly it.
+*
+*    The cleanup consists of two parts, pruning and simplification. Pruning works
+*    by calculating the closest distance reached by the line segment between any
+*    two pairs of sequential points, and cuts out anything between two points
+*    when their line segments get close. This algorithm will never compare two
+*    consecutive line segments. Obviously the segments (p1,p2) and (p2,p3) will
+*    get very close (they touch), but there would be nothing to trim between them.
+*    The simplification step uses the Ramer-Douglas-Peucker algorithm.
+*    See Wikipedia for description.
+*
+*    The simplification and pruning algorithms do not alter the path in memory,
+*    and are designed to run in the background. They each have a parameter that
+*    decides how long they will run before saving their state and returning. They
+*    are both anytime algorithms, which is helpful when memory is filling up and
+*    we just need to quickly identify a handful of points which can be deleted.
+*    The algorithms can also report if they are done. If not, the thorough cleanup
+*    will just have to wait for a bit until they are both done.
+*/
+
+AC_SafeRTL::AC_SafeRTL(const AP_AHRS& ahrs) :
+    _ahrs(ahrs),
+    _active(false),
+    _simplification_stack_last_index(-1),
+    _prunable_loops_last_index(-1)
+{
+    AP_Param::setup_object_defaults(this, var_info);
+    _simplification_bitmask = std::bitset<SAFERTL_PATH_LEN_MAX>().set(); //initialize to 0b1111...
+    _time_of_last_good_position = AP_HAL::millis();
+}
+
+// initialise safe rtl including setting up background processes
+void AC_SafeRTL::init()
+{
+    // check if user has disabled SafeRTL
+    if (_desired_path_len == 0 || is_zero(_accuracy)) {
+        _active = false;
+        return;
+    }
+
+    // protect against repeated call to init
+    if (_initialised) {
+        return;
+    }
+
+    // constrain the path length, in case the user decided to make the path unreasonably long.
+    _desired_path_len = MIN(SAFERTL_PATH_LEN_MAX, _desired_path_len);
+
+    // allocate arrays
+    _path = (Vector3f*)malloc(_desired_path_len * sizeof(Vector3f));
+    _prunable_loops = (loop*)malloc(_desired_path_len * sizeof(loop));
+    _simplification_stack = (start_finish*)malloc(_desired_path_len * sizeof(start_finish));
+
+    // check if memory allocation failed
+    if (_path == nullptr || _prunable_loops == nullptr || _simplification_stack == nullptr) {
+        log_action(SRTL_DEACTIVATED_INIT_FAILED);
+        gcs().send_text(MAV_SEVERITY_WARNING, "SafeRTL: unavailable init failed");
+        free(_path);
+        free(_prunable_loops);
+        free(_simplification_stack);
+        _initialised = false;
+        return;
+    } else {
+        _current_path_len = _desired_path_len;
+        _initialised = true;
+    }
+
```

Comment:
I might skip telling the user.  I see dataflash does that but what if every subsystem did that.  Most users won't know what to do with the info.